### PR TITLE
Make testdriver click/send_keys/actions work in more browsing contexts

### PR DESCRIPTION
**Pull request by jgraham**
This is a slightly hacky change to allow testdriver actions to work in
more browsing contexts. The basic idea is that when we want to run an
action in a different borsing context, we pass wptrunner a context
list like [window_id, frame_id...], and use that to select the correct
browsing context just for the duration of the action, before switching
back.

There are a number of limitations with the current patch, some more
serious than others:

* So far this is only implmented for testdriver commands that take an
explicit element. Other commands could be modified to also allow
passing in an explicit context.

* It must be possible to get a window reference and set the name
property (see below). This means this approach only works for
windows that are same origin with the test window.

* WebDriver implementations don't generally support returning a window
object from script, so we can't return a window handle
directly. Instead we either use the existing window.name property or
add a window.name when we try to start the action. Then in wptrunner
we do a linear search of all windows to find the one with the
appropriate name. We have to use window.name rather than writing a
custom property into the window global because the way marionette
sandboxes js we aren't able to read the global property.

However despite these limitations this makes the feature considerably
more versatile than it was previously.

### CODE REVIEW

File: tools/wptrunner/wptrunner/testdriver-extra.js
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/testdriver-extra.js:53]: I'd call this `currentWindow` or `currentNode` (if we lean into describing it as a tree-walk), as `target` to me is actually `elementWindow` (as that's what we're trying to find).

File: tools/wptrunner/wptrunner/executors/base.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/executors/base.py:840]: Can we get here? I don't quite follow how `window` (`self.context[0]`) can be falsy in the code as written; isn't it always set if context is non-null?

(Or is there a situation where `self.protocol.base.current_window` could have been `None`?)

**Review by stephenmcgruer — COMMENTED** [review]
Thanks for doing this, I think it's a great improvement to testdriver!

Looked at the non-test files so far.

File: infrastructure/testdriver/click_iframe.html
Comment (stephenmcgruer) [inline infrastructure/testdriver/click_iframe.html:12]: Optional, but this (and the other tests) could be written as a promise test which might make the failure case clearer:

```
promise_test(async t => {
  await new Promise(resolve => { addEventListener("load", resolve); });
  let child = frames[0];
  let button = child.document.getElementById("button");
  await test_driver.click(button);
  assert_equals(child.document.getElementById("log").textContent, "PASS");
});
```

For me its a cleaner syntax (but that's obviously subjective), but it also reduces the error handling required. (Thrown errors or rejected promises should get a full stack trace).

File: infrastructure/testdriver/click_nested.html
Comment (stephenmcgruer) [inline infrastructure/testdriver/click_nested.html:14]: I'd suggest a comment (or modifying the test title element) to explain that we're making sure that click reaches the correct target in the case of multiple windows/frames.

File: infrastructure/testdriver/click_iframe.html
Comment (stephenmcgruer) [inline infrastructure/testdriver/click_iframe.html:15]: Is the iframe definitely loaded at this point?

File: infrastructure/testdriver/click_iframe.html
Comment (stephenmcgruer) [inline infrastructure/testdriver/click_iframe.html:13]: I notice all of the tests have this outer load listener - is it necessary? (I.e. what are we waiting for?)

**Review by stephenmcgruer — COMMENTED** [review]
Generally LGTM modulo comments, but seems like there are some failures in the unittests (unable to find the target window?) and the lint (mostly trailing-whitespace, one usage of console).

File: infrastructure/testdriver/click_iframe.html
Comment (jgraham) [reply infrastructure/testdriver/click_iframe.html:13]: For the iframes to load

File: infrastructure/testdriver/click_iframe.html
Comment (jgraham) [reply infrastructure/testdriver/click_iframe.html:15]: Yes, the load event on the document is blocked on iframes loading.

Comment (stephenmcgruer) [reply infrastructure/testdriver/click_iframe.html:15]: TIL! Thanks :)

File: infrastructure/testdriver/click_iframe.html
Comment (jgraham) [reply infrastructure/testdriver/click_iframe.html:12]: I made everything single page tests instead which also removes most of the boilerplate. idk which is better overall but I'd say they're at least comparable.

File: tools/wptrunner/wptrunner/testdriver-extra.js
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/testdriver-extra.js:143]:
```suggestion
                        context = action_context;
```

This is causing:

```
Error: Actions must be in a single context
    at Object.window.test_driver_internal.action_sequence (http://web-platform.test:8000/resources/testdriver.js:739:35)
    at Object.action_sequence (http://web-platform.test:8000/resources/testdriver.js:185:48)
    at Actions.send (http://web-platform.test:8000/resources/testdriver-actions.js:68:26)
    at Test.<anonymous> (http://web-platform.test:8000/infrastructure/testdriver/actions/elementTiming.html:52:6)
    at Test.step (http://web-platform.test:8000/resources/testharness.js:1977:25)
    at async_test (http://web-platform.test:8000/resources/testharness.js:571:22)
    at http://web-platform.test:8000/infrastructure/testdriver/actions/elementTiming.html:36:1
```

**Review by stephenmcgruer — APPROVED** [review]
LGTM after the infra tests are fixed.

File: infrastructure/testdriver/click_iframe.html
Comment (Hexcles) [inline infrastructure/testdriver/click_iframe.html:20]:
```suggestion
            done();
```

File: infrastructure/testdriver/click_nested.html
Comment (Hexcles) [inline infrastructure/testdriver/click_nested.html:26]:
```suggestion
                done();
```

File: infrastructure/testdriver/click_window.html
Comment (Hexcles) [inline infrastructure/testdriver/click_window.html:19]:
```suggestion
                done();
```

File: resources/testdriver.js
Comment (Hexcles) [inline resources/testdriver.js:132]: This doesn't match the updated documentation.

**Review by Hexcles — COMMENTED** [review]
Shall we revert https://github.com/web-platform-tests/wpt/commit/a9d6d1b058f9d37cabaf735ea60f9b87acc8a02b#diff-8f0dc082cb16d568172a0e6f39f0045c as well?

**Pull request merged**
